util/network-devp2p/src/discovery.rs

```diff
@@ -541,10 +561,22 @@ impl<'a> Discovery<'a> {
 		};
 
 		if let Some((node, ping_reason)) = expected_node {
-			if let PingReason::FromDiscoveryRequest(target) = ping_reason {
+			if let PingReason::FromDiscoveryRequest(target, validity) = ping_reason {
 				self.respond_with_discovery(target, &node)?;
+				// kirushik: I would prefer to probe the network id of the remote node here, and add it to the nodes list if it's on "our" net --
+				// but `on_packet` happens synchronously, so doing the full TCP handshake ceremony here is a bad idea.
+				// So instead we just LRU-caching most recently seen nodes to avoid unnecessary pinging
+				match validity {
+					NodeValidity::UnknownNode | NodeValidity::ExpiredNode(NodeCategory::Observed) => {
```

Comment:
I guess we should update `observed_at` and `endpoint` in case of `ValidNode`s for completeness as well, no?